### PR TITLE
Country History Improvements

### PR DESCRIPTION
**Pull request by BrickPi**
Adds simple `PROPERTY_RW` macro for getters & setters.
Adds `CountryInstance` class to represent in-game countries.
Adds function to apply country history to country instances.

### CODE REVIEW

File: src/openvic-simulation/country/CountryInstance.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        CountryParty const* PROPERTY_RW(ruling_party)
+        Date PROPERTY_RW(last_election)
+        decimal_map_t<Ideology const*> PROPERTY(upper_house)
+        Province const* PROPERTY_RW(capital)
+        GovernmentType const* PROPERTY_RW(government_type)
+        fixed_point_t PROPERTY_RW(plurality)
+        NationalValue const* PROPERTY_RW(national_value)
+        bool PROPERTY_RW(civilised)
+        fixed_point_t PROPERTY_RW(prestige)
+        std::vector<Reform const*> PROPERTY(reforms)
```
Comment (Hop311) [inline src/openvic-simulation/country/CountryInstance.hpp:25]: This works for now as a place to put the country history reforms, eventually we'll need a map from reform groups to single active reforms from that group.

File: src/openvic-simulation/utility/Getters.hpp [hunk 2]
```diff
@@ -63,4 +63,13 @@ public: \
...
 private:
 
+#define PROPERTY_RW(NAME) \
+	PROPERTY(NAME) \
+public: \
+	void set_##NAME(decltype(NAME) new_##NAME) { \
+		NAME = new_##NAME; \
+	} \
+\
+private:
```
Comment (Hop311) [inline src/openvic-simulation/utility/Getters.hpp:73]: Good job with how you've used this macro, only with things which we're happy to pass by value (so primitive types and basic primitive type wrappers like `Date` and `fixed_point_t`). As for special logic to choose the argument type and control how it's assigned, let's see what Spartan can come up with (but that can be a future PR).

**Review by Hop311 — COMMENTED** [review]
Good work with the macro and `CountryInstance`, it's gonna be one of the most important structs in the game so I'm glad to see it finally added.
*As a note for the future*, I notice you're often having to check to avoid suplicates in vectors - the solution would be `std::set` except that doesn't preserve the order elements get added in so we'll need something else, could just be a wrapper around `std::vector` with a special insert function but I'm sure Spartan knows some useful libraries we could take a look at.

**Comment by Hop311** [comment]
Also you used `PROPERTY_RW` in the first commit and only added the macro itself in the second one, then the third is a tiny change - might as well squash it all into a single commit.

**Comment by Hop311** [comment]
Also thinking again, some things in CountryInstance should be moved to Country, e.g. cores are tied to the tag, when s

**Comment by BrickPi** [comment]
> Also thinking again, some things in CountryInstance should be moved to Country, e.g. cores are tied to the tag, when switching tag, eg via a decision, you have to replace the old tag's cores with the new ones yourself as part of the effect script. Others I'm not sure about and may need testing, it's possible there are even things that stay with the tag _and_ instance upon tag switch.

I was thinking we should keep `Country` just static values. Cores are currently stored by province which is where I think they fit best.

**Review by Hop311 — APPROVED** [review]
If we can keep Country as a static definition that would be ideal, especially for being able to reset and apply a different bookmark without needing to reload the starting state. We'll need to do some testing, eg switch to a new tag then release the original tag again and see what data it retains.

**Pull request merged**